[inductor][cpp] improve vector contiguous checks for FloorDiv and ModularIndexing

jgong5 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #117263
* #117260
* __->__ #117221

Fix https://github.com/pytorch/pytorch/issues/114488

The PR tries to enable contiguous vector loads for cases where we can reduce `FloorDiv` and `ModularIndexing` in the vectorized loop.

Take the index expression in test case `test_vec_contiguous_ModularIndexing` for example.
`14336*x0 + 256*x1 + 128*((x2//256)) + ModularIndexing(x2, 1, 128) + 7168*ModularIndexing(x2, 128, 2)` can be reduced to `14336*x0 + 256*x1 + x2 + 128*x2_div_c0 + 7168*x2_mod_c0 + x2_mod_c1` where `x2` is a vectorized loop variable and the vector length is 16. This means we can do vectorized load for this index. Check the code comment for more details:
https://github.com/pytorch/pytorch/pull/117221/files#diff-5ab7b0235e2076a5fc6629ba0b109208940f5b94f5c13babc3e0f87cf4fcec82R317-R329

cc @voznesenskym @penguinwu @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

Commit:
[inductor][cpp] improve vector contiguous checks for FloorDiv and ModularIndexing

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] improve vector contiguous checks for FloorDiv and ModularIndexing"


Fix https://github.com/pytorch/pytorch/issues/114488

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -311,6 +311,60 @@ def stride_at(var: sympy.Symbol, index: sympy.Expr):
+    floor_divs: Dict[sympy.Expr, Dict[sympy.Expr, sympy.Expr]] = {}
+    collect(index, FloorDiv(var, div), floor_divs, lambda m: m[div] % vec_length == 0)
+    for i, floor_div in enumerate(floor_divs):
+        freevar = sympy.Symbol(f"{var}_div_c{i}")
```
Where is the defined in the output code?  It looks like we just create the var then never define it.

jgong5 replied:
See comment in https://github.com/pytorch/pytorch/pull/117221#discussion_r1449746217

jansel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -311,6 +311,60 @@ def stride_at(var: sympy.Symbol, index: sympy.Expr):
+        or (m[div] == 1 and m[mod] % vec_length == 0),
+    )
+    for i, modular in enumerate(modulars):
+        freevar = sympy.Symbol(f"{var}_mod_c{i}")
```
Where is the defined in the output code?  It looks like we just create the var then never define it.

jgong5 replied:
This simplified index is not used for codegen but for reasoning about whether we can do contiguous vector load/store in a vectorized loop. The reason why these free symbols are added is to make sure the simplified index expression is sound, i.e., for any given loop itervar value assigned in the vectorized loop, there exist assignments of these free symbols so that the simplified index expression value equals to the original index expression and these free symbols are irrelevant to the loop itervar in the vector range. For example, for the index expression below
`14336*x0 + 256*x1 + 128*((x2//256)) + ModularIndexing(x2, 1, 128) + 7168*ModularIndexing(x2, 128, 2)`
It is simplified to
`14336*x0 + 256*x1 + x2 + 128*x2_div_c0 + 7168*x2_mod_c0 + x2_mod_c1`
given `x2` is a vectorized loop var.
For any assignment for `x2` where `x2 = a*16 + b`, `b` stands for the vector range from 0 to 15, we have corresponding assignments of `x2_div_c0`, `x2_mod_c0` and `x2_mod_c1` that make the simplified index expression. For example, take `a = 17` and `b = 3`, we have `x2_div_c0 = 1`, `x2_mod_c0 = 2` and `x2_mod_c1 = -256`, which don't depend on `b`. With that, we know it is safe to do vectorized load/store with this index expression.

jansel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -311,6 +311,60 @@ def stride_at(var: sympy.Symbol, index: sympy.Expr):
+       variable when `mod` is divisible by 16.
+    """
+
+    def collect(expr, pattern, collected, filter=None):
```
This function seems rather similar to `expr.replace`.  For example:
https://github.com/pytorch/pytorch/blob/04604eea8ad201777a11c80c9dc64112e21cfb3c/torch/_inductor/sizevars.py#L120-L152

jgong5 replied:
Thanks for the suggestion. I revised the code accordingly.

Commit:
Update on "[inductor][cpp] improve vector contiguous checks for FloorDiv and ModularIndexing"


Fix https://github.com/pytorch/pytorch/issues/114488

The PR tries to enable contiguous vector loads for cases where we can reduce `FloorDiv` and `ModularIndexing` in the vectorized loop.

Take the index expression in test case `test_vec_contiguous_ModularIndexing` for example.
`14336*x0 + 256*x1 + 128*((x2//256)) + ModularIndexing(x2, 1, 128) + 7168*ModularIndexing(x2, 128, 2)` can be reduced to `14336*x0 + 256*x1 + x2 + 128*x2_div_c0 + 7168*x2_mod_c0 + x2_mod_c1` where `x2` is a vectorized loop variable and the vector length is 16. This means we can do vectorized load for this index. Check the code comment for more details:
https://github.com/pytorch/pytorch/pull/117221/files#diff-5ab7b0235e2076a5fc6629ba0b109208940f5b94f5c13babc3e0f87cf4fcec82R317-R329

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] improve vector contiguous checks for FloorDiv and ModularIndexing"


Fix https://github.com/pytorch/pytorch/issues/114488

The PR tries to enable contiguous vector loads for cases where we can reduce `FloorDiv` and `ModularIndexing` in the vectorized loop.

Take the index expression in test case `test_vec_contiguous_ModularIndexing` for example.
`14336*x0 + 256*x1 + 128*((x2//256)) + ModularIndexing(x2, 1, 128) + 7168*ModularIndexing(x2, 128, 2)` can be reduced to `14336*x0 + 256*x1 + x2 + 128*x2_div_c0 + 7168*x2_mod_c0 + x2_mod_c1` where `x2` is a vectorized loop variable and the vector length is 16. This means we can do vectorized load for this index. Check the code comment for more details:
https://github.com/pytorch/pytorch/pull/117221/files#diff-5ab7b0235e2076a5fc6629ba0b109208940f5b94f5c13babc3e0f87cf4fcec82R317-R329

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] improve vector contiguous checks for FloorDiv and ModularIndexing"


Fix https://github.com/pytorch/pytorch/issues/114488

The PR tries to enable contiguous vector loads for cases where we can reduce `FloorDiv` and `ModularIndexing` in the vectorized loop.

Take the index expression in test case `test_vec_contiguous_ModularIndexing` for example.
`14336*x0 + 256*x1 + 128*((x2//256)) + ModularIndexing(x2, 1, 128) + 7168*ModularIndexing(x2, 128, 2)` can be reduced to `14336*x0 + 256*x1 + x2 + 128*x2_div_c0 + 7168*x2_mod_c0 + x2_mod_c1` where `x2` is a vectorized loop variable and the vector length is 16. This means we can do vectorized load for this index. Check the code comment for more details:
https://github.com/pytorch/pytorch/pull/117221/files#diff-5ab7b0235e2076a5fc6629ba0b109208940f5b94f5c13babc3e0f87cf4fcec82R317-R329

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -311,6 +311,61 @@ def stride_at(var: sympy.Symbol, index: sympy.Expr):
+    original_index = index
+
+    div = sympy.Wild("divisor")
+    index = index.replace(FloorDiv(var, div), visit_indexing_div)
```
```suggestion
    if index.has(FloorDiv):
       index = index.replace(FloorDiv(var, div), visit_indexing_div)
```
might be a slight compile time improvement

jgong5 replied:
Done.

jansel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -311,6 +311,61 @@ def stride_at(var: sympy.Symbol, index: sympy.Expr):
+    index = index.replace(FloorDiv(var, div), visit_indexing_div)
+
+    mod = sympy.Wild("modulus")
+    index = index.replace(ModularIndexing(var, div, mod), visit_modular_indexing)
```
same as above

jgong5 replied:
Done.

jansel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -311,6 +311,61 @@ def stride_at(var: sympy.Symbol, index: sympy.Expr):
+       when `div` is divisible by 16.
+    2. `ModularIndexing(x3, 1, mod)` can be simplified to `x3 + c` where `c` is a free
+       variable when `mod` is divisible by 16.
+    """
```
Add a comment explaining how this is only used for analysis not codegen

jgong5 replied:
Comments added.

Commit:
Update on "[inductor][cpp] improve vector contiguous checks for FloorDiv and ModularIndexing"


Fix https://github.com/pytorch/pytorch/issues/114488

The PR tries to enable contiguous vector loads for cases where we can reduce `FloorDiv` and `ModularIndexing` in the vectorized loop.

Take the index expression in test case `test_vec_contiguous_ModularIndexing` for example.
`14336*x0 + 256*x1 + 128*((x2//256)) + ModularIndexing(x2, 1, 128) + 7168*ModularIndexing(x2, 128, 2)` can be reduced to `14336*x0 + 256*x1 + x2 + 128*x2_div_c0 + 7168*x2_mod_c0 + x2_mod_c1` where `x2` is a vectorized loop variable and the vector length is 16. This means we can do vectorized load for this index. Check the code comment for more details:
https://github.com/pytorch/pytorch/pull/117221/files#diff-5ab7b0235e2076a5fc6629ba0b109208940f5b94f5c13babc3e0f87cf4fcec82R317-R329

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jgong5 commented:
@pytorchbot merge
